subprojects/core/src/main/java/org/gradle/api/internal/collections/DefaultPendingSource.java

```diff
@@ -58,7 +59,11 @@ private void realize(Iterable<TypedCollector<T>> collectors) {
             if (flushAction != null) {
                 pending.remove(collector);
                 ImmutableList.Builder<T> builder = ImmutableList.builder();
-                collector.collectInto(builder);
+                ValueSupplier.Value<Void> valueWithSideEffects = collector.collectInto(builder);
+                if (!valueWithSideEffects.isMissing()) {
+                    // run side effects if any
+                    valueWithSideEffects.get();
+                }
```

Comment:
Similarly to `AbstractIterationOrderRetainingElementSource`, we should discuss whether it is a reasonable thing to do here